src/Microsoft.Extensions.Hosting.Abstractions/BackgroundService.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Hosting
+{
+    /// <summary>
+    /// Base class for implementing a long running <see cref="IHostedService"/>.
+    /// </summary>
+    public abstract class BackgroundService : IHostedService
+    {
+        private Task _executingTask;
+        private CancellationTokenSource _cts;
+
+        /// <summary>
+        /// This method is called when the <see cref="IHostedService"/> starts. The implementation should return a task that represents
+        /// the lifetime of the long running operation(s) being performed.
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> that fires when the <see cref="IHostedService"/> is stopped or cancelled.</param>
+        /// <returns>A <see cref="Task"/> that represents the long running operations.</returns>
+        protected abstract Task ExecuteAsync(CancellationToken cancellationToken);
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            // If the token is cancellable, create a linked token so we can trigger cancellation 
+            // on shutdown and based on the incoming token
+            _cts = cancellationToken.CanBeCanceled ?
+                CancellationTokenSource.CreateLinkedTokenSource(cancellationToken) :
```

Comment:
Linking these is odd, their lifetimes are completely different. I understand why you have both, but maybe you should pass them in separately to ExecuteAsync.

Reply:
@Tratcher passing two tokens to Execute async would be insane. What would the code do besides link them? Either we're saying that you cant cancel Startup of the service or we can.

Either way passing 2 tokens is a bad design.